README.md
# @gregoranders/jhu-covid19

<!-- markdownlint-disable-next-line MD033 -->
## [<img src="./typescript.svg" />][typescript-url]

<!-- markdownlint-disable-next-line MD013 -->
## Provides access to the data published at the [COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University](https://github.com/CSSEGISandData/COVID-19)

## [API Docs](./docs/api/index.md)

[![License][license-image]][license-url]
[![Issues][issues-image]][issues-url]
[![Code maintainability][code-maintainability-image]][code-maintainability-url]
[![Code issues][code-issues-image]][code-issues-url]
[![Code Technical Debt][code-tech-debt-image]][code-tech-debt-url]

[![types][npm-types-image]][npm-types-url]
[![node][node-image]][node-url]

[![Main Language][language-image]][code-metric-url]
[![Languages][languages-image]][code-metric-url]
[![Code Size][code-size-image]][code-metric-url]
[![Repository Size][repo-size-image]][code-metric-url]

## Requires `fetch`

## Features

- [TypeScript][typescript-url]
- [Jest][jest-url] Unit Tests with Code Coverage
- [API Docs Generation][api-url]
- GitHub CI Integration (feature, development, master, release)
- Code Quality via [Code Climate](./docs/codeclimate.md) and [Codacy](./docs/codacy.md)

| GitHub                                                           | Coveralls                                                                  |                                                                              |
| :--------------------------------------------------------------- | :------------------------------------------------------------------------- | :--------------------------------------------------------------------------- |
| [![Release Build][release-image]][release-url]                   |                                                                            | [![npm][npm-image]][npm-url]                                                 |
| [![Master Build][master-build-image]][master-url]                | [![Master Coverage][master-coveralls-image]][master-coveralls-url]         | [![Master Version][master-version-image]][master-version-url]                |
| [![Development Build][development-build-image]][development-url] | [![Test Coverage][development-coveralls-image]][development-coveralls-url] | [![Development Version][development-version-image]][development-version-url] |

## Example

```sh
npm install @gregoranders/jhu-covid19
```

```ts
import fetch from 'node-fetch';
import Provider from '@gregoranders/jhu-covid19';

const main = async () => {
  const provider = new Provider(fetch);
  const model = await provider.get();

  console.log(model);
};

main();
```

### Clone repository

```sh
git clone https://github.com/gregoranders/ts-jhu-covid19
```

### Install dependencies

```sh
npm install
```

### Build

```sh
npm run build
```

### Testing

#### Test using [Jest][jest-url]

```sh
npm test
```

### Code Climate Checks [docker required](docs/codeclimate.md)

```sh
npm run codeclimate
```

### Clear

```sh
npm run clear
```

[release-url]: https://github.com/gregoranders/ts-jhu-covid19/releases
[master-url]: https://github.com/gregoranders/ts-jhu-covid19/tree/master
[development-url]: https://github.com/gregoranders/ts-jhu-covid19/tree/development
[code-metric-url]: https://github.com/gregoranders/ts-jhu-covid19/search?l=TypeScript
[license-url]: https://github.com/gregoranders/ts-jhu-covid19/blob/master/LICENSE
[license-image]: https://img.shields.io/github/license/gregoranders/ts-jhu-covid19.svg
[master-version-url]: https://github.com/gregoranders/ts-jhu-covid19/blob/master/package.json
[master-version-image]: https://img.shields.io/github/package-json/v/gregoranders/ts-jhu-covid19/master
[development-version-url]: https://github.com/gregoranders/ts-jhu-covid19/blob/development/package.json
[development-version-image]: https://img.shields.io/github/package-json/v/gregoranders/ts-jhu-covid19/development
[issues-url]: https://github.com/gregoranders/ts-jhu-covid19/issues
[issues-image]: https://img.shields.io/github/issues-raw/gregoranders/ts-jhu-covid19.svg
[master-build-image]: https://github.com/gregoranders/ts-jhu-covid19/workflows/Master%20CI/badge.svg
[development-build-image]: https://github.com/gregoranders/ts-jhu-covid19/workflows/Development%20CI/badge.svg
[master-coveralls-url]: https://coveralls.io/github/gregoranders/ts-jhu-covid19?branch=master
[master-coveralls-image]: https://img.shields.io/coveralls/github/gregoranders/ts-jhu-covid19/master
[development-coveralls-image]: https://img.shields.io/coveralls/github/gregoranders/ts-jhu-covid19/development
[development-coveralls-url]: https://coveralls.io/github/gregoranders/ts-jhu-covid19?branch=development
[code-maintainability-url]: https://codeclimate.com/github/gregoranders/ts-jhu-covid19/maintainability
[code-maintainability-image]: https://img.shields.io/codeclimate/maintainability/gregoranders/ts-jhu-covid19
[code-issues-url]: https://codeclimate.com/github/gregoranders/ts-jhu-covid19/maintainability
[code-issues-image]: https://img.shields.io/codeclimate/issues/gregoranders/ts-jhu-covid19
[code-tech-debt-url]: https://codeclimate.com/github/gregoranders/ts-jhu-covid19/maintainability
[code-tech-debt-image]: https://img.shields.io/codeclimate/tech-debt/gregoranders/ts-jhu-covid19
[npm-url]: https://www.npmjs.com/package/@gregoranders/jhu-covid19
[npm-image]: https://img.shields.io/npm/v/@gregoranders/jhu-covid19
[node-url]: https://www.npmjs.com/package/@gregoranders/jhu-covid19
[node-image]: https://img.shields.io/node/v/@gregoranders/jhu-covid19
[npm-types-url]: https://www.npmjs.com/package/@gregoranders/jhu-covid19
[npm-types-image]: https://img.shields.io/npm/types/@gregoranders/jhu-covid19
[release-image]: https://img.shields.io/github/release/gregoranders/ts-jhu-covid19
[language-image]: https://img.shields.io/github/languages/top/gregoranders/ts-csv
[languages-image]: https://img.shields.io/github/languages/count/gregoranders/ts-csv
[code-size-image]: https://img.shields.io/github/languages/code-size/gregoranders/ts-csv
[repo-size-image]: https://img.shields.io/github/repo-size/gregoranders/ts-csv
[typescript-url]: http://www.typescriptlang.org/
[jest-url]: https://jestjs.io
[api-url]: https://api-extractor.com/


CHANGELOG.md
## [0.0.10](https://github.com/gregoranders/ts-jhu-covid19/compare/v0.0.9...v0.0.10) (2023-10-28)

#### 0.0.8 (2020-06-28)

##### Build System / Dependencies

- **workflows:**
  - minor changes to workflows ([7f4523fb](https://github.com/gregoranders/ts-jhu-covid19/commit/7f4523fb6fb0e8d3b3ea3f17e24578e66d52dee8))
  - add release workflow ([8bf5421d](https://github.com/gregoranders/ts-jhu-covid19/commit/8bf5421db5e3355651d0a0ac279c8a51a3e3f8e2))
  - refactor ([6e97f43f](https://github.com/gregoranders/ts-jhu-covid19/commit/6e97f43f9fbf1db76b8001da6150fff59bf5bea8))
- **dependencies:** dependencies check weekly ([4b7b788a](https://github.com/gregoranders/ts-jhu-covid19/commit/4b7b788aa7e6ab4c19fe44687f335fa033a1ccca))
- **cc:** run code doverage on demand ([d3eea22a](https://github.com/gregoranders/ts-jhu-covid19/commit/d3eea22ac8c615c4f71cbb99e6fc2bc3aa75e74e))

##### Chores

- **release:**
  - prepare 0.0.8 Release ([a09431c2](https://github.com/gregoranders/ts-jhu-covid19/commit/a09431c2b74f7349af97f47903bb4b757e7aa3ad))
  - 0.0.7 release ([a24f0a39](https://github.com/gregoranders/ts-jhu-covid19/commit/a24f0a39e86396f9cdafd9ee44fbd111f1575696))
  - exclude files from package ([0a3722d1](https://github.com/gregoranders/ts-jhu-covid19/commit/0a3722d19d5e046c0f68fa7fd009f02567c27245))
  - remove unused file ([f127916b](https://github.com/gregoranders/ts-jhu-covid19/commit/f127916b5b0449f59cc7c06a04327fbf9393b299))
- **dependencies:** update dependencies ([3bc03f8a](https://github.com/gregoranders/ts-jhu-covid19/commit/3bc03f8a1acaa62fc0eccae40e3a68af6ba790dc))
- **package:** update dependencies ([be0b092e](https://github.com/gregoranders/ts-jhu-covid19/commit/be0b092e9e913ea3df133796e57f1b254a8a0e50))

##### Documentation Changes

- **README:**
  - add typescript logo ([29cb1bff](https://github.com/gregoranders/ts-jhu-covid19/commit/29cb1bff02509e3ddb91b9a1d42b547482808f27))
  - add requirenment on fetch ([913d5d9c](https://github.com/gregoranders/ts-jhu-covid19/commit/913d5d9cc4ae67e2633922a38e482df84d815920))
  - update docs ([c553786d](https://github.com/gregoranders/ts-jhu-covid19/commit/c553786d143cffe1c516929c1de4b81414067cc9))
- **GITHUB:** add pull request template ([62e4b053](https://github.com/gregoranders/ts-jhu-covid19/commit/62e4b0535d6338cbb8d1106a3d2c4d50f79c78d9))
- **CONTRIB:** add CONTRIBUTING.md ([26fbf687](https://github.com/gregoranders/ts-jhu-covid19/commit/26fbf687c96701415dc5b605b391232d8c254a60))
- **api:** refactor api docs ([6822d35f](https://github.com/gregoranders/ts-jhu-covid19/commit/6822d35feb40d5f87e032405b21105886ea3c446))

##### New Features

- **code quality:** refactor eslint integration ([7ee9829c](https://github.com/gregoranders/ts-jhu-covid19/commit/7ee9829c3e1849719c7eb3da226b8ebd12c3de20))

##### Bug Fixes

- **fetch:** add fetch abstraction layer ([9c2492aa](https://github.com/gregoranders/ts-jhu-covid19/commit/9c2492aa8a7ce2c34ee167fea46d5ba466a1ba35))
- **eslint:** fix eslint integration ([2ce1134a](https://github.com/gregoranders/ts-jhu-covid19/commit/2ce1134aa20c004456152394784e0614eb6c8379))
- **vscode:** fix version lense plugin ([00607411](https://github.com/gregoranders/ts-jhu-covid19/commit/0060741193933c7766b6f31eee77bd70bace8c80))

##### Other Changes

- bump @microsoft/api-documenter from 7.8.14 to 7.8.17 ([900c9a84](https://github.com/gregoranders/ts-jhu-covid19/commit/900c9a844477197114a1c9fec342c6c37d40e390))
- bump @microsoft/api-extractor from 7.8.12 to 7.8.15 ([3a0d047b](https://github.com/gregoranders/ts-jhu-covid19/commit/3a0d047beda0c1f0043342a6e3440a411665c4b1))
- bump @types/jest from 26.0.0 to 26.0.3 ([eec03085](https://github.com/gregoranders/ts-jhu-covid19/commit/eec0308544e2088fc120ebcb19853c65859677c9))
- bump @microsoft/api-extractor from 7.8.12 to 7.8.14 ([71aeeb55](https://github.com/gregoranders/ts-jhu-covid19/commit/71aeeb554de0759cdb3fc5b4c309f5b109b0b9a9))
- bump @types/node from 14.0.13 to 14.0.14 ([ca86bea8](https://github.com/gregoranders/ts-jhu-covid19/commit/ca86bea84a872327ee8e4f7a07e46dc36cbbf350))
- bump jest from 26.0.1 to 26.1.0 ([00e71dd1](https://github.com/gregoranders/ts-jhu-covid19/commit/00e71dd12676910182cacc917d10402c888196f9))
- bump jest-cli from 26.0.1 to 26.1.0 ([b1573485](https://github.com/gregoranders/ts-jhu-covid19/commit/b1573485abdb3eb846a7e9c3792b6dd995b5e00e))
- bump @typescript-eslint/eslint-plugin from 3.3.0 to 3.4.0 ([5bc4823a](https://github.com/gregoranders/ts-jhu-covid19/commit/5bc4823ae56d498f43b03918f7daa45499e32525))
- bump @typescript-eslint/parser from 3.3.0 to 3.4.0 ([47f3d172](https://github.com/gregoranders/ts-jhu-covid19/commit/47f3d1728969a280f161acefeb497f58e8d9262e))
- bump eslint from 7.3.0 to 7.3.1 ([8f34d047](https://github.com/gregoranders/ts-jhu-covid19/commit/8f34d047a3c89905429717108ca9a27ab8adc49c))
- bump jest-junit from 11.0.0 to 11.0.1 ([ce4fcde2](https://github.com/gregoranders/ts-jhu-covid19/commit/ce4fcde252045b491cc1c6b8c650d6b5dc64037e))
- bump ts-jest from 26.1.0 to 26.1.1 ([435893b4](https://github.com/gregoranders/ts-jhu-covid19/commit/435893b4f10738832bbf487564011e002700ba7f))
- bump @microsoft/api-documenter from 7.8.13 to 7.8.14 ([75c7c6a6](https://github.com/gregoranders/ts-jhu-covid19/commit/75c7c6a6d1e073ffadf05891eff5c0e2748bbb85))
- bump jest-junit from 10.0.0 to 11.0.0 ([d9ced382](https://github.com/gregoranders/ts-jhu-covid19/commit/d9ced3823afeeb80573021e91629968c68413c02))
- bump lint-staged from 10.2.10 to 10.2.11 ([660d9f1f](https://github.com/gregoranders/ts-jhu-covid19/commit/660d9f1f4bb99ef73ee46bf729f3db048c9fca12))
- bump jest-spec-reporter from 1.0.13 to 1.0.14 ([6b72f948](https://github.com/gregoranders/ts-jhu-covid19/commit/6b72f948ccec985bdf1d8a83c22cdf4094a32a4d))
- bump @microsoft/api-documenter from 7.8.12 to 7.8.13 ([fedfbecf](https://github.com/gregoranders/ts-jhu-covid19/commit/fedfbecfe408785fa7e61c80b96b2ad11e05b47b))

src/index.spec.ts
import * as TestSubject from '.';

describe(`${TestSubject.libname} ${TestSubject.libversion}`, () => {
  describe('exports', () => {
    it('libname', () => {
      expect(TestSubject.libname).toBe('@gregoranders/jhu-covid19');
    });

    it('libversion', () => {
      expect(TestSubject.libversion).toBe('0.0.11');
    });

    it('liburl', () => {
      expect(TestSubject.liburl).toMatch(/^https/);
    });

    it('default', () => {
      expect(TestSubject.default).toBeDefined();
    });

    it('Provider', () => {
      expect(TestSubject.Provider).toBeDefined();
    });

    it('Provider = default', () => {
      expect(TestSubject.Provider).toStrictEqual(TestSubject.default);
    });
  });
});


src/provider.spec.ts
import { libname, libversion } from '.';
import * as TestSubject from './provider';

const resolved = (text: string) =>
  Promise.resolve({
    text: () => Promise.resolve(text),
  });

const mockFetch = (lookup: string, confirmed: string, deaths: string, recovered: string) =>
  jest.fn().mockImplementation((url: string) => {
    switch (url) {
      case 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/UID_ISO_FIPS_LookUp_Table.csv': {
        return resolved(lookup);
      }
      case 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv': {
        return resolved(confirmed);
      }
      case 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv': {
        return resolved(deaths);
      }
      case 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_recovered_global.csv': {
        return resolved(recovered);
      }
      default: {
        console.error(url);
        return { text: () => Promise.reject(url) };
      }
    }
  });

describe(`${libname} ${libversion} - provider`, () => {
  describe('exports', () => {
    it('default', () => {
      expect(TestSubject.default).toBeDefined();
    });

    it('Provider', () => {
      expect(TestSubject.Provider).toBeDefined();
    });

    it('Provider  = default', () => {
      expect(TestSubject.Provider).toBe(TestSubject.default);
    });
  });

  describe('collect', () => {
    it('empty on empty responses', async () => {
      const mockedFetch = mockFetch('', '', '', '');

      const provider = new TestSubject.Provider(mockedFetch);
      const data = await provider.get();
      expect(mockedFetch.mock.calls[0][1].headers).toStrictEqual({
        'Accept-Encoding': 'br, gzip, deflate',
      });
      expect(mockedFetch).toHaveBeenCalledTimes(4);
      expect(data).toHaveLength(0);
    });

    it('metrics', async () => {
      let lookup = 'UID,iso2,iso3,code3,FIPS,Admin2,Province_State,Country_Region,Lat,Long_,Combined_Key,Population\n';
      lookup += '276,DE,DEU,276,,,,Germany,51.165691,10.451526,Germany,83783945\n';
      lookup +=
        '3601,AU,AUS,36,,,Australian Capital Territory,Australia,-35.4735,149.0124,"Australian Capital Territory, Australia",0';

      const header = 'Province/State,Country/Region,Lat,Long,1/1/20,2/1/20\n';
      const germany = ',Germany,51.0,9.0,';
      const australia = 'Australian Capital Territory,Australia,-35.4735,149.0124,';

      let confirmed = header;
      confirmed += germany;
      confirmed += [0, 1].join(',');
      confirmed += '\n';
      confirmed += australia;
      confirmed += [1, 3].join(',');
      confirmed += '\n';

      let deaths = header;
      deaths += germany;
      deaths += [0, 0].join(',');
      deaths += '\n';
      deaths += australia;
      deaths += [0, 1].join(',');
      deaths += '\n';

      let recovered = header;
      recovered += germany;
      recovered += [0, 0].join(',');
      recovered += '\n';
      recovered += australia;
      recovered += [0, 1].join(',');
      recovered += '\n';

      const mockedFetch = mockFetch(lookup, confirmed, deaths, recovered);

      const provider = new TestSubject.Provider(mockedFetch);
      const data = await provider.get();
      expect(mockedFetch.mock.calls[0][1].headers).toStrictEqual({
        'Accept-Encoding': 'br, gzip, deflate',
      });

      expect(mockedFetch).toHaveBeenCalledTimes(4);

      expect(data).toHaveLength(2);

      expect(data[0].country).toStrictEqual('Australia');
      expect(data[0].state).toStrictEqual('Australian Capital Territory');
      expect(data[0].metrics).toHaveLength(2);
      expect(data[0].metrics[1].confirmed).toStrictEqual(3);
      expect(data[0].metrics[1].active).toStrictEqual(1);
      expect(data[0].metrics[1].dead).toStrictEqual(1);
      expect(data[0].metrics[1].recovered).toStrictEqual(1);
      expect(data[0].metrics[1].diff.confirmed).toStrictEqual(2);
      expect(data[0].metrics[1].diff.active).toStrictEqual(0);
      expect(data[0].metrics[1].diff.dead).toStrictEqual(1);
      expect(data[0].metrics[1].diff.recovered).toStrictEqual(1);
      expect(data[0].metrics[1].avrg[5].confirmed).toStrictEqual(2 / 5);
      expect(data[0].metrics[1].avrg[7].confirmed).toStrictEqual(2 / 7);
      expect(data[0].metrics[1].avrg[14].confirmed).toStrictEqual(2 / 14);
      expect(data[0].metrics[1].avrg[21].confirmed).toStrictEqual(2 / 21);
      expect(data[0].metrics[1].avrg[28].confirmed).toStrictEqual(2 / 28);

      expect(data[1].country).toStrictEqual('Germany');
      expect(data[1].state).toBeUndefined();
      expect(data[1].metrics).toHaveLength(2);
      expect(data[1].metrics[1].confirmed).toStrictEqual(1);
      expect(data[1].metrics[1].active).toStrictEqual(1);
      expect(data[1].metrics[1].dead).toStrictEqual(0);
      expect(data[1].metrics[1].recovered).toStrictEqual(0);
      expect(data[1].metrics[1].diff.confirmed).toStrictEqual(1);
      expect(data[1].metrics[1].diff.active).toStrictEqual(1);
      expect(data[1].metrics[1].diff.dead).toStrictEqual(0);
      expect(data[1].metrics[1].diff.recovered).toStrictEqual(0);
      expect(data[1].metrics[1].avrg[5].dead).toStrictEqual(0);
      expect(data[1].metrics[1].avrg[7].dead).toStrictEqual(0);
      expect(data[1].metrics[1].avrg[14].dead).toStrictEqual(0);
      expect(data[1].metrics[1].avrg[21].dead).toStrictEqual(0);
      expect(data[1].metrics[1].avrg[28].dead).toStrictEqual(0);
    });
  });
});


src/index.ts
/**
 * Provides access to the data published at the COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University.
 *
 * https://github.com/CSSEGISandData/COVID-19
 *
 * @packageDocumentation
 */

/**
 * library name
 *
 * @public
 * @readonly
 */
export const libname = '@gregoranders/jhu-covid19';

/**
 * library version
 *
 * @public
 * @readonly
 */
export const libversion = '0.0.11';

/**
 * library homepage
 *
 * @public
 * @readonly
 */
export const liburl = 'https://gregoranders.github.io/ts-jhu-covid19/';

/**
 * @public
 */
export {
  FetchLike,
  FetchLikeOptions,
  FetchLikeResult,
  Metric,
  MetricValue,
  MetricValueAvrg,
  MetricValueAvrgType,
  MetricValueType,
  Model,
  Provider,
  default,
} from './provider';

/**
 * @public
 */


src/provider.ts
import { FetchLike, ModelCollector, RowModel, RowModelValue } from './collector';

/**
 * metric value type
 *
 * @example
 * ```ts
 * {
 *   confirmed: 1,
 *   dead: 1,
 *   recovered: 1,
 *   active: 1,
 * }
 * ```
 * {@link MetricValue | MetricValue}
 *
 * @public
 * @readonly
 */
export type MetricValueType = 'confirmed' | 'dead' | 'recovered' | 'active';

/**
 * metric value
 *
 * @example
 * ```ts
 * {
 *   confirmed: 1,
 *   dead: 1,
 *   recovered: 1,
 *   active: 1,
 * }
 * ```
 * {@link MetricValueType | MetricValueType}
 *
 * @public
 * @readonly
 */
export interface MetricValue extends Record<MetricValueType, number> {
  /**
   * confirmed cases
   *
   * @public
   * @readonly
   */
  confirmed: number;
  /**
   * deaths
   *
   * @public
   * @readonly
   */
  dead: number;
  /**
   * recovered cases
   *
   * @public
   * @readonly
   */
  recovered: number;
  /**
   * active cases
   *
   * @remarks
   * `active` is calculated (`confirmed` - `dead` - `recovered`),
   * some data sets do not provide a recovered value
   *
   * @public
   * @readonly
   */
  active: number;
}

/**
 * precalculated average type - in days
 *
 * @example
 * ```ts
 * {
 *    5: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *    7: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   14: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   21: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   28: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 * }
 * ```
 * {@link MetricValueAvrg | MetricValueAvrg}
 *
 * @public
 * @readonly
 */
export type MetricValueAvrgType = 5 | 7 | 14 | 21 | 28;

/**
 * metric value averages
 *
 * @example
 * ```ts
 * {
 *    5: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *    7: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   14: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   21: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 *   28: { confirmed: 1, dead: 1, recovered: 1, active: 1 }
 * }
 * ```
 * {@link MetricValueAvrgType | MetricValueAvrgType}
 *
 * @public
 * @readonly
 */
export interface MetricValueAvrg extends Record<MetricValueAvrgType, MetricValue> {
  /**
   * 5 days average
   *
   * @public
   * @readonly
   */
  5: MetricValue;
  /**
   * 7 days average
   *
   * @public
   * @readonly
   */
  7: MetricValue;
  /**
   * 17 days average
   *
   * @public
   * @readonly
   */
  14: MetricValue;
  /**
   * 21 days average
   *
   * @public
   * @readonly
   */
  21: MetricValue;
  /**
   * 28 days average
   *
   * @public
   * @readonly
   */
  28: MetricValue;
}

/**
 * daily metric
 *
 * @public
 * @readonly
 */
export interface Metric extends MetricValue {
  /**
   * difference to day before
   *
   * @public
   * @readonly
   */
  diff: MetricValue;
  /**
   * average values
   *
   * @public
   * @readonly
   */
  avrg: MetricValueAvrg;
  /**
   * ratio according to population
   *
   * @public
   * @readonly
   */
  ratio: MetricValue & {
    /**
     * difference to day before
     *
     * @public
     * @readonly
     */
    diff: MetricValue;
    /**
     * ratio according to population
     *
     * @public
     * @readonly
     */
    avrg: MetricValueAvrg;
  };
  /**
   * UNIX UTC timestamp
   *
   * @public
   * @readonly
   */
  timestamp: number;
}

/**
 * country/region
 *
 * @public
 */
export interface Model {
  /**
   * country
   *
   * @public
   * @readonly
   */
  country: string;
  /**
   * state/province
   *
   * @public
   * @readonly
   */
  state?: string | null;
  /**
   * population
   *
   * @public
   * @readonly
   */
  population: number;
  /**
   * series of daily mertics
   *
   * @public
   * @readonly
   */
  metrics: Metric[];
}

/**
 * @internal
 */
const DEFAULT_METRIC_VALUE = {
  confirmed: 0,
  dead: 0,
  recovered: 0,
  active: 0,
} as MetricValue;

/**
 * @internal
 */
export class ModelProcessor {
  private _model: Model[] = [];

  public constructor(model: readonly RowModel[]) {
    if (model && model.length > 0) {
      this._model = this._process(model);
    }
  }

  public get model(): Model[] {
    return this._model;
  }

  private _process(models: readonly RowModel[]) {
    return models.map((model) => {
      const factor = model.population ? model.population / 100_000 : 1;
      const mapped = {
        country: model.country,
        population: model.population,
      } as Model;

      if (model.state && model.state.length > 0) {
        mapped.state = model.state;
      }

      mapped.metrics = this._metrics(model, factor);

      return mapped;
    });
  }

  private _metrics(model: RowModel, factor: number) {
    return model.values.map((value, index, all) => {
      const avrg = this._averages(all, index);
      const diff = this._diff(value, index, all);
      const base = {
        confirmed: value.confirmed,
        dead: value.deaths,
        recovered: value.recovered,
        active: this._active(value),
      };
      return {
        ...base,
        diff,
        avrg,
        ratio: {
          ...this._ratio(base, factor),
          diff: this._ratio(diff, factor),
          avrg: this._averagesRatio(avrg, factor),
        },
        timestamp: value.timestamp,
      } as Metric;
    });
  }

  private _diff(value: RowModelValue, index: number, all: RowModelValue[]): MetricValue {
    return index
      ? {
          confirmed: value.confirmed - all[index - 1].confirmed,
          dead: value.deaths - all[index - 1].deaths,
          recovered: value.recovered - all[index - 1].recovered,
          active: this._active(value) - this._active(all[index - 1]),
        }
      : { ...DEFAULT_METRIC_VALUE };
  }

  private _averages(values: RowModelValue[], index: number): MetricValueAvrg {
    return {
      5: this._avrg(5, index, values),
      7: this._avrg(7, index, values),
      14: this._avrg(14, index, values),
      21: this._avrg(21, index, values),
      28: this._avrg(28, index, values),
    };
  }

  private _averagesRatio(values: MetricValueAvrg, factor: number): MetricValueAvrg {
    return {
      5: this._ratio(values[5], factor),
      7: this._ratio(values[7], factor),
      14: this._ratio(values[14], factor),
      21: this._ratio(values[21], factor),
      28: this._ratio(values[28], factor),
    };
  }

  private _avrg(back: number, index: number, all: RowModelValue[]): MetricValue {
    const sum = { ...DEFAULT_METRIC_VALUE };

    if (index === 0) {
      sum.confirmed = all[index].confirmed;
      sum.dead = all[index].deaths;
      sum.recovered = all[index].recovered;
      sum.active = this._active(all[index]);
    } else {
      for (let index_ = index; index_ > index - back && index_ > 0; index_--) {
        const diff = this._diff(all[index_], index_, all);
        sum.confirmed += diff.confirmed;
        sum.dead += diff.dead;
        sum.recovered += diff.recovered;
        sum.active += diff.active;
      }
    }

    return this._ratio(sum, back);
  }

  private _active(value: RowModelValue) {
    return value.confirmed - value.deaths - value.recovered;
  }

  private _ratio(value: MetricValue, factor: number): MetricValue {
    return {
      confirmed: this._safeDiv(value.confirmed, factor),
      dead: this._safeDiv(value.dead, factor),
      recovered: this._safeDiv(value.recovered, factor),
      active: this._safeDiv(value.active, factor),
    };
  }

  private _safeDiv(value: number, factor: number) {
    if (value === 0) {
      return 0;
    }
    return value / factor;
  }
}

/**
 * data provider
 *
 * @remarks
 * provides access to the data hosted at {@link https://github.com/CSSEGISandData/COVID-19}
 *
 * @example
 * ```ts
 * import fetch from 'node-fetch';
 * import Provider from '@gregoranders/jhu-covid19';
 *
 * const main = async () => {
 *   const provider = new Provider(fetch);
 *   const model = await provider.get();
 *
 *   console.log(model);
 * };
 *
 * main();
 * ```
 *
 * @public
 */
export class Provider {
  /**
   * constructor
   *
   * @param _fetch - `fetch` like interface implementation
   *
   * @example
   * ```ts
   *   import fetch from 'node-fetch';
   *
   *   const provider = new Provider(fetch);
   * ```
   *
   * @public
   */
  public constructor(private _fetch: FetchLike) {}

  /**
   * Fetches the JHU datset, aggregates, precalculates values, diffs and averages.
   *
   * @remarks
   * fetches data over the network
   *
   * @returns the aggregated covid19 data provided by the JHU
   *
   * @public
   * @readonly
   */
  public async get(): Promise<Model[]> {
    const collector = new ModelCollector(this._fetch);
    const model = await collector.collect();
    const processor = new ModelProcessor(model);
    return processor.model;
  }
}

/**
 * Default export
 *
 * @public
 */
export default Provider;

export { FetchLike, FetchLikeOptions, FetchLikeResult } from './collector';
